### PR TITLE
Add options for initial convolution, including channel value offset invariance, for ResNet

### PR DESCRIPTION
**Pull request by nickwp**
The main feature of this PR is to introduce the capability for time-shift invariance of ResNet CNNs.

For this, the following features have been added.

ResNet model:
- Options `first_kernel_size` and `first_stride` to choose the network's first convolution's kernel size and stride (both default to the current  fixed value of 1 because that's what's typically used for mPMT data)
- Option `shift_inv_channels` to list a subset of channel indices to be made invariant an arbitrary global offset, with the intention is that this is the hit time channel(s) of the data, so that a shift in the global time offset does not affect the output of the network
- Use explicit padding function before the initial convolution, instead of the built in padding of the convolution function, to allow custom modified padding (explained below) and asymmetric padding size used when the initial convolution's kernel size is even.

CNN (and CNN mPMT) datasets:
- Add option `use_median_unhit_times` to use the median hit time to set the values of the time channel for pixels that correspond to either a location where no PMT exists or to an unhit PMT (currently and now by default these pixels' time values are set to zero, and charge values are still always set to zero for unhit or non-existent PMTs)

Time translation invariance is achieved by replacing the initial convolution with a modified convolution that constrains to zero the sum weights that multiply values from the specified channels. This constraint is enforced by subtracting the mean value of that subset of weights from each of the respective weights. This is done for all the chosen channels together, assuming that they would all be offset by the same value and we want invariance to this global offset.

There are two cases that can break this invariance if not dealt with appropriately:
- The time values of unhit PMTs or non-existent PMTs are currently set to zero by default. This would allow the network to learn the difference between a hit PMT time and an unhit/non-existent neighbouring PMT's zero time value, which equals the hit's absolute time
- When the padding mode is `"zeros"`, the initial tensor is padded with zeros before the first convolution to ensure the convolution's output tensor has the same spatial dimensions as the input, and these zero values create the same problem as mentioned above

In both cases, the solution is to use a value that shifts along with any global shifts to actual hit times. The median hit time value was chosen, since this is less likely than the mean to be skewed by hit time outliers that may occur in real data but not in training data. For unhit/non-existent PMT times, the new option in the CNN dataset enables this. For padding, in the case where zero padding is used along with shift invariance, the padding is modified to use the values of the pixel in the corner of the image at (0,0), which generally does not correspond to an actual, so the time value of a non-existent PMT is replicated for the padding. In the case of using the double-cover transformation, the (0,0) pixel does have an existing PMT that might have a genuine hit time, but when using double-cover the padding should be set to circular padding anyway, not zero padding.

The shift-invariance setup also requires that the initial convolution has a stride that is less than the kernel size (and so the kernel size must be at least 2x2) otherwise there is no overlap of the receptive field of two neighbouring locations and the network, being only able to learn differences between times within a receptive field, would be unable to learn relative differences between values between one receptive field and the neighbouring receptive field. To ensure there is overlap, the initial kernel size must be at least 2 and the initial stride less than the initial kernel size.

I tested with the HK FD setup for electron position regression, specifying that the time channel `0` should be shift invariant and the following updated config parameters:
```
dataset:
  use_median_unhit_times: true
model:
  conv_pad_mode: zeros
  first_kernel_size: 4
  first_stride: 2
  shift_inv_channels: [0]
```
The accuracy slightly exceeded the same configuration without time-shift invariance (but hyperparameters may not have been fully optimised anyway)
```T-shift invariant:  Average evaluation position error: 20.22```
```         Baseline:  Average evaluation position error: 23.16```
Testing the trained model on data that had a 1000 ns offset applied to all hit times, showed only negligible differences likely due to floating-point rounding differences:
```
Tested on unshifted times:
Step 0/1393555 Evaluation position error: 22.291, loss: 82.814
Step 10/1393555 Evaluation position error: 10.459, loss: 18.231
Step 20/1393555 Evaluation position error: 12.684, loss: 26.813
Tested on 1000 ns shifted times:
Step 0/1393555 Evaluation position error: 22.294, loss: 82.833
Step 10/1393555 Evaluation position error: 10.461, loss: 18.238
Step 20/1393555 Evaluation position error: 12.689, loss: 26.837
```

Note about mPMT data:
This setup may not work well on mPMT data, since this usually starts with a 1x1 convolution over each mPMT, which cannot overlap with a neighbouring mPMT. A time-shift invariant network in this case could learn relative hit time differences between hits within each mPMT, but not across neighbouring mPMTs. Changing the initial convolution to 2x2 with a  stride of 1 should work. If this does not learn well, another possibility to implement might be to retain the initial 1x1 convolution, add back in the mean hit time at each mPMT manually as an additional channel in the second layer, which would be a second convolution with kernel size >1 and shift invariance applied to this manually added mean hit time channel.

### CODE REVIEW

**Comment by nickwp** [comment]
Using this code version, I have re-run training of the HK FD and compared directly to Andrew's results.
With an initial convolution size 4, stride 2, with time invariance enabled on the time channel, and using median hit time for the "zeros":
```
data:
  dataset:
    use_median_unhit_times: true
model:
  first_kernel_size: 4
  first_stride: 2
  shift_inv_channels: [0]
```
and even using ResNet-50 instead of ResNet-152 to get training to converge quickly enough, the performance is notably improved.
image

So I think this is enough to merge now
